### ISSUE TITLE
Thoughts on a loop() feature?

### ISSUE THREAD

**Issue by aidan-casey** [Anteris-Dev/autotask-client#7]
Wondering about adding a `loop()` method to the QueryBuilder. Essentially instead of having to run `paginate()` and create your own loop over the next pages, you could pass in a callback and have it automatically loop. This could be useful for instances where you have a large number of records to retrieve and iterate over and you want to condense the code.

Example:

```php

$query = $client->contacts()->query()->where('firstName', 'eq', 'Aidan');

$query->loop(function ($contact) {

    echo $contact->firstName;

});

```

**Comment by aidan-casey** [comment]
I am removing this from the v0.2.0 milestone as I do not feel urgent about the feature. That said, if people would like to see this implemented, I do not see a problem with doing so.

**Comment by Darrrren** [comment]
Yeah I agree, perhaps not a priority at the moment, we can always foreach through the results:

```php
$query = $client->contacts()->query()->where('firstName', 'eq', 'Aidan')->get();

foreach($query as $contact) 
{
    echo $contact->firstName . PHP_EOL;
}
```

**Comment by aidan-casey** [comment]
Thanks for the feedback! Definitely can foreach, the problem becomes if the records that Autotask has is greater than 500, it won't return all of them in one request so you would have to paginate the results. This would look like this:

```php
$page = $client->contacts()->query()->where('firstName', 'eq', 'Aidan')->paginate();

while (true) {
    
    // Loop through the results and echo the first name of every contact
    foreach ($page->collection as $contact) {
         echo $contact->firstName . PHP_EOL;
    }

    // If there is no next page, stop looping
    if (! $page->hasNextPage()) {
        break;
    }
    
    // Set the page variable to the results of the next page
    $page = $page->nextPage();

}
```

So if we ever implement a loop feature, the problem it would solve is that situation where you say to yourself, I don't care how many records there are, I want to take all of them and perform this action for each of them.

**Comment by Darrrren** [comment]
Yes the 500 limit is a pain, love the ->loop() idea now I understand it better. I would love to see this in 0.3.0 but to be fair, we have paginate() and the code you've illustrated above so it's not a show stopper just now. Nice job!

**Comment by aidan-casey** [comment]
Thanks @Darrrren! I am going to continue to shelve this for 0.3.0, but I am adding it to the 0.4.0 milestone. 0.3.0 is going to bring a lot to the table with automated tests from #1 (this will make actual requests against the Autotask API and ensure every response gives the intended results). I would like to continue focus on bringing this functionality so we can assure anyone using this package that the 160+ supported routes work when they need to use them.

**Comment by Darrrren** [comment]
Sounds great. I totally agree, let's get some solid foundations in place first and then use that to build on in the future. I've had to accept an all day meeting today so I'll clone it tonight / tomorrow and take a first look as soon as I can for you.

**Comment by aidan-casey** [comment]
Closed by #17.
